File: projects/ar-editor-preview/apps/ios/ARCompanion/Transport/LiveKitStreamer.swift
// apps/ios/ARCompanion/Transport/LiveKitStreamer.swift
// ARKit データを LiveKit 経由で Unity Editor へ送信する。

import Foundation
import LiveKit
import ARKit

// ─────────────────────────────────────────────────────────────────
// 接続設定
// ─────────────────────────────────────────────────────────────────

struct StreamerConfig {
    var serverUrl:   String = "ws://192.168.1.100:7880"
    var roomName:    String = "ar-preview"
    var identity:    String = "ios-device"
    var token:       String
    var videoWidth:  Int    = 1280
    var videoHeight: Int    = 720
    var videoFps:    Int    = 30
}

// ─────────────────────────────────────────────────────────────────
// LiveKitStreamer
// ─────────────────────────────────────────────────────────────────

@MainActor
final class LiveKitStreamer: ObservableObject {

    private var room: Room?
    private var cameraTrack: LocalVideoTrack?

    @Published var isConnected = false
    @Published var connectionState: ConnectionState = .disconnected

    // ─── 接続 ───────────────────────────────────────────────────

    func connect(config: StreamerConfig) async throws {
        let r = Room()
        room = r

        // 接続オプション
        let connectOptions = ConnectOptions(
            autoSubscribe: false    // Publisher は Subscribe 不要
        )

        // 映像パブリッシュオプション
        let videoOptions = VideoPublishOptions(
            name     : "ar-camera",
            encoding : VideoEncoding(
                maxBitrate: 2_000_000,
                maxFps    : config.videoFps
            ),
            simulcast: false
        )

        try await r.connect(
            config.serverUrl,
            config.token,
            connectOptions: connectOptions
        )

        isConnected    = true
        connectionState = .connected

        // カメラトラック作成・パブリッシュ
        let cameraCapturer = CameraCapturer(options: CameraCapturerOptions(
            position   : .back,
            preferredPixelFormat: .kCVPixelFormatType_420YpCbCr8BiPlanarFullRange
        ))

        cameraTrack = LocalVideoTrack.createCameraTrack(capturer: cameraCapturer)
        try await r.localParticipant.publishVideoTrack(
            track: cameraTrack!,
            publishOptions: videoOptions
        )

        // セッション状態送信
        try await sendSessionState()
    }

    func disconnect() async {
        await room?.disconnect()
        room           = nil
        isConnected    = false
        connectionState = .disconnected
    }

    // ─── ARFrame 送信 ────────────────────────────────────────────

    func sendFrame(_ data: ARFrameData) {
        guard isConnected, let participant = room?.localParticipant else { return }

        let envelope = Arpreview_V1_Envelope.with {
            $0.frame = Arpreview_V1_ARFrame.with {
                $0.timestampMs = data.timestampMs
                $0.frameNumber = data.frameNumber
                $0.pose = Arpreview_V1_CameraPose.with {
                    $0.position = Arpreview_V1_Vec3.with { $0.x = data.tx; $0.y = data.ty; $0.z = data.tz }
                    $0.rotation = Arpreview_V1_Quaternion.with { $0.x = data.rx; $0.y = data.ry; $0.z = data.rz; $0.w = data.rw }
                }
                $0.intrinsics = Arpreview_V1_CameraIntrinsics.with {
                    $0.fx = data.fx; $0.fy = data.fy
                    $0.ppx = data.ppx; $0.ppy = data.ppy
                    $0.w = data.imageWidth; $0.h = data.imageHeight
                }
                $0.light = Arpreview_V1_LightEstimate.with {
                    $0.averageIntensity  = data.averageIntensity
                    $0.colorTemperature  = data.colorTemperature
                }
            }
        }

        Task {
            guard let bytes = try? envelope.serializedData() else { return }
            try? await participant.publishData(
                data        : bytes,
                reliability : .lossy      // フレームは最新優先
            )
        }
    }

    // ─── プレーン更新送信 ────────────────────────────────────────

    func sendPlanes(_ planes: [ARPlaneInfo]) {
        guard isConnected, !planes.isEmpty,
              let participant = room?.localParticipant else { return }

        let protoPlanes: [Arpreview_V1_PlaneData] = planes.map { p in
            Arpreview_V1_PlaneData.with {
                $0.id    = p.id.uuidString
                $0.event = {
                    switch p.event {
                    case .added:   return .trackingEventAdded
                    case .updated: return .trackingEventUpdated
                    case .removed: return .trackingEventRemoved
                    }
                }()
                $0.alignment = {
                    switch p.alignment {
                    case .horizontal: return .planeAlignmentHorizontalUp
                    case .vertical:   return .planeAlignmentVertical
                    @unknown default: return .planeAlignmentNotAxisAligned
                    }
                }()
                $0.center   = Arpreview_V1_Vec3.with { $0.x = p.center.x; $0.y = p.center.y; $0.z = p.center.z }
                $0.rotation = Arpreview_V1_Quaternion.with {
                    $0.x = p.rotation.vector.x; $0.y = p.rotation.vector.y
                    $0.z = p.rotation.vector.z; $0.w = p.rotation.vector.w
                }
                $0.extents      = Arpreview_V1_Vec2.with { $0.x = p.extentX; $0.y = p.extentZ }
                $0.boundaryXz   = p.boundaryXZ
            }
        }

        let envelope = Arpreview_V1_Envelope.with {
            $0.planes = Arpreview_V1_ARPlaneUpdate.with {
                $0.timestampMs = Int64(Date().timeIntervalSince1970 * 1000)
                $0.planes      = protoPlanes
            }
        }

        Task {
            guard let bytes = try? envelope.serializedData() else { return }
            try? await participant.publishData(
                data        : bytes,
                reliability : .reliable    // プレーンは確実に届ける
            )
        }
    }

    // ─── セッション状態送信 ──────────────────────────────────────

    private func sendSessionState() async throws {
        guard let participant = room?.localParticipant else { return }

        let envelope = Arpreview_V1_Envelope.with {
            $0.session = Arpreview_V1_SessionState.with {
                $0.timestampMs = Int64(Date().timeIntervalSince1970 * 1000)
                $0.status      = .sessionStatusReady
                $0.platform    = .platformIos
                $0.deviceModel = UIDevice.current.model
                $0.osVersion   = UIDevice.current.systemVersion
                $0.arSdkVersion = Bundle.main.object(forInfoDictionaryKey: "ARKitVersion") as? String ?? ""
                $0.features    = ["planes", "light_estimate"]
            }
        }

        guard let bytes = try? envelope.serializedData() else { return }
        try await participant.publishData(data: bytes, reliability: .reliable)
    }
}

File: projects/ar-editor-preview/apps/android/app/src/main/java/com/arpreview/transport/LiveKitStreamer.kt
// apps/android/app/src/main/java/com/arpreview/transport/LiveKitStreamer.kt
// ARCore データを LiveKit 経由で Unity Editor へ送信する。
// Video Track (カメラ映像) + Data Channel (Protobuf メタデータ) の 2 チャネル。

package com.arpreview.transport

import android.content.Context
import android.util.Log
import com.arpreview.ar.ARFrameData
import com.arpreview.ar.ARPlaneData
import com.arpreview.ar.PlaneEventType
import com.arpreview.proto.*
import io.livekit.android.LiveKit
import io.livekit.android.room.Room
import io.livekit.android.room.participant.LocalParticipant
import io.livekit.android.room.track.CameraPosition
import io.livekit.android.room.track.LocalVideoTrack
import io.livekit.android.room.track.LocalVideoTrackOptions
import io.livekit.android.room.track.video.Camera2Capturer
import io.livekit.android.events.RoomEvent
import io.livekit.android.events.collect
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*

private const val TAG = "LiveKitStreamer"

// ─────────────────────────────────────────────────────────────────
// 接続設定
// ─────────────────────────────────────────────────────────────────

data class StreamerConfig(
    val serverUrl: String = "ws://192.168.1.100:7880",
    val roomName:  String = "ar-preview",
    val identity:  String = "android-device",
    val token:     String,             // LiveKit JWT トークン
    val videoWidth:  Int = 1280,
    val videoHeight: Int = 720,
    val videoFps:    Int = 30,
)

// ─────────────────────────────────────────────────────────────────
// LiveKitStreamer
// ─────────────────────────────────────────────────────────────────

class LiveKitStreamer(private val context: Context) {

    private val scope       = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var room: Room? = null
    private var videoTrack: LocalVideoTrack? = null

    // 接続状態
    val isConnected: Boolean get() =
        room?.state == Room.State.CONNECTED

    // ─── 接続 ───────────────────────────────────────────────────

    suspend fun connect(config: StreamerConfig): Result<Unit> = runCatching {
        val lkRoom = LiveKit.create(context).apply {
            room = this
        }

        // ルームイベント監視
        scope.launch {
            lkRoom.events.collect { event ->
                when (event) {
                    is RoomEvent.Connected    -> Log.i(TAG, "✅ Connected to ${config.roomName}")
                    is RoomEvent.Disconnected -> Log.w(TAG, "❌ Disconnected: ${event.error}")
                    is RoomEvent.Reconnecting -> Log.i(TAG, "🔄 Reconnecting...")
                    else -> {}
                }
            }
        }

        lkRoom.connect(config.serverUrl, config.token)

        // カメラ映像パブリッシュ
        publishVideo(lkRoom.localParticipant, config)

        // セッション状態を送信
        sendSessionState(lkRoom.localParticipant, config.identity)
    }

    suspend fun disconnect() {
        videoTrack?.stop()
        videoTrack = null
        room?.disconnect()
        room = null
    }

    fun destroy() {
        scope.cancel()
    }

    // ─── 映像パブリッシュ ────────────────────────────────────────

    private suspend fun publishVideo(participant: LocalParticipant, config: StreamerConfig) {
        val track = participant.createVideoTrack(
            name    = "ar-camera",
            options = LocalVideoTrackOptions(
                position = CameraPosition.BACK,
                captureParams = VideoCaptureParameter(
                    width     = config.videoWidth,
                    height    = config.videoHeight,
                    maxFps    = config.videoFps,
                )
            )
        )
        track.startCapture()
        participant.publishVideoTrack(track)
        videoTrack = track
        Log.i(TAG, "Video track published: ${config.videoWidth}x${config.videoHeight}@${config.videoFps}fps")
    }

    // ─── ARFrame 送信 ────────────────────────────────────────────

    fun sendFrame(data: ARFrameData) {
        val participant = room?.localParticipant ?: return

        val envelope = Envelope(
            frame = ARFrame(
                timestamp_ms = data.timestampMs,
                frame_number = data.frameNumber.toULong(),
                pose = CameraPose(
                    position = Vec3(x = data.tx, y = data.ty, z = data.tz),
                    rotation = Quaternion(x = data.rx, y = data.ry, z = data.rz, w = data.rw),
                ),
                intrinsics = CameraIntrinsics(
                    fx  = data.fx,  fy  = data.fy,
                    ppx = data.ppx, ppy = data.ppy,
                    w   = data.imageWidth,
                    h   = data.imageHeight,
                ),
                light = LightEstimate(
                    average_intensity  = data.averageLuminance,
                    color_temperature  = data.colorTemperature,
                ),
            )
        )

        scope.launch {
            try {
                participant.publishData(
                    data      = envelope.encode(),
                    reliability = DataPublishReliability.LOSSY,  // フレームデータは最新優先
                )
            } catch (e: Exception) {
                Log.w(TAG, "sendFrame error: ${e.message}")
            }
        }
    }

    // ─── プレーン更新送信 ────────────────────────────────────────

    fun sendPlanes(planes: List<ARPlaneData>) {
        if (planes.isEmpty()) return
        val participant = room?.localParticipant ?: return

        val protoPlanes = planes.map { p ->
            PlaneData(
                id             = p.id,
                event          = when (p.event) {
                    PlaneEventType.ADDED   -> TrackingEventType.TRACKING_EVENT_ADDED
                    PlaneEventType.UPDATED -> TrackingEventType.TRACKING_EVENT_UPDATED
                    PlaneEventType.REMOVED -> TrackingEventType.TRACKING_EVENT_REMOVED
                },
                tracking_state = TrackingState.TRACKING_STATE_TRACKING,
                alignment      = when (p.alignment.ordinal) {
                    0 -> PlaneAlignment.PLANE_ALIGNMENT_HORIZONTAL_UP
                    1 -> PlaneAlignment.PLANE_ALIGNMENT_HORIZONTAL_DOWN
                    2 -> PlaneAlignment.PLANE_ALIGNMENT_VERTICAL
                    else -> PlaneAlignment.PLANE_ALIGNMENT_NOT_AXIS_ALIGNED
                },
                center         = Vec3(x = p.cx, y = p.cy, z = p.cz),
                rotation       = Quaternion(x = p.rx, y = p.ry, z = p.rz, w = p.rw),
                extents        = Vec2(x = p.extentX, y = p.extentZ),
                boundary_xz    = p.boundaryXZ.toList(),
                subsumed_by_id = p.subsumedById ?: "",
            )
        }

        val envelope = Envelope(
            planes = ARPlaneUpdate(
                timestamp_ms = System.currentTimeMillis(),
                planes       = protoPlanes,
            )
        )

        scope.launch {
            try {
                participant.publishData(
                    data        = envelope.encode(),
                    reliability = DataPublishReliability.RELIABLE,  // プレーンは確実に届ける
                )
            } catch (e: Exception) {
                Log.w(TAG, "sendPlanes error: ${e.message}")
            }
        }
    }

    // ─── セッション状態送信 ──────────────────────────────────────

    private suspend fun sendSessionState(participant: LocalParticipant, identity: String) {
        val envelope = Envelope(
            session = SessionState(
                timestamp_ms   = System.currentTimeMillis(),
                status         = SessionStatus.SESSION_STATUS_READY,
                platform       = Platform.PLATFORM_ANDROID,
                device_model   = android.os.Build.MODEL,
                os_version     = android.os.Build.VERSION.RELEASE,
                ar_sdk_version = com.google.ar.core.ArCoreApk.getInstance()
                    .checkAvailability(context).name,
                features       = listOf("planes", "light_estimate"),
            )
        )

        try {
            participant.publishData(
                data        = envelope.encode(),
                reliability = DataPublishReliability.RELIABLE,
            )
            Log.i(TAG, "Session state sent")
        } catch (e: Exception) {
            Log.w(TAG, "sendSessionState error: ${e.message}")
        }
    }
}
